app/src/main/java/ru/spbhse/pocketmagic/NetworkController.java

```diff
@@ -0,0 +1,81 @@
+package ru.spbhse.pocketmagic;
+
+import android.app.AlertDialog;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.support.v7.app.AppCompatActivity;
+
+import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
+
+import java.nio.ByteBuffer;
+
+public class NetworkController {
+
+    private static AppCompatActivity currentActivity;
+    private static Controller currentController;
+    private static Network network;
+
+    public static Network createNetwork(GoogleSignInAccount account) {
+        network = new Network(account);
+        return network;
+    }
+
+    public static void startGame() {
+        Intent intent = new Intent(currentActivity, LoadActivity.class);
+        intent.putExtra("GameType", GameType.MULTIPLAYER);
+        currentActivity.startActivity(intent);
+        currentActivity.finish();
+    }
+
+    public static void finishGame() {
+        if (currentActivity.getClass() == GameActivity.class) {
+            ((GameActivity) currentActivity).finishGame(GameResult.ERROR);
+        }
+    }
+
+    synchronized public static void setUI(AppCompatActivity activity) {
+        currentActivity = activity;
+        if (currentActivity != null && currentActivity.getClass() == GameActivity.class) {
+            currentController = ((GameActivity) currentActivity).getController();
+        } else {
+            currentController = null;
+        }
+    }
+
+    synchronized public static Context getContext() {
+        return currentActivity.getApplicationContext();
```

Comment:
The problem with this synchronized method is that it doesn't provide any synchronization guarantees about its result. Returned ApplicationContext could be accessed without synchronization while setUI can change the currentActivity, making the ApplicationContext obsolete. Can multithreaded code be avoided here? In general, it's a good idea to run away from multithreading whenever possible.